_drafts/ember-orbit.md

```diff
@@ -0,0 +1,477 @@
+
+```js
+import { RequestStrategy } from '@orbit/coordinator';
+
+export default {
+  create() {
+    return new RequestStrategy({
+      name: 'remote-update',
+
+      source: 'store',
+      on: 'beforeUpdate',
+
+      target: 'remote',
+      action: 'push',
+
+      blocking: true,
+    });
+  }
+};
+```
+
+This says that before every update to the store, we should push the corresponding changes to the remote store. “Blocking” means that the remote push must succeed before the local change is allowed to proceed; we’ll learn more about this option later.
+
+For strategies to run, we need to enable Orbit’s coordinator. It's created for us when our app starts, but not activated automatically; we'll see why later. Let's activate it in our application route:
+
+`app/routes/application.js`:
+
+```js
+import Route from '@ember/routing/route';
+import { inject as service } from '@ember/service';
+
+export default Route.extend({
+  dataCoordinator: service(),
+
+  beforeModel() {
+    const coordinator = this.dataCoordinator;
+    return coordinator.activate();
+  },
+});
+```
+
+Now if we run our app and check the network tab of our browser, every time we add a message we'll see an outgoing POST request to our API, and a 201 Created response:
+
+SCREENSHOT
+
+## EventLoggingStrategy: Console Output
+
+To see what’s happening in more detail, let's add event logging so we can see it in the browser console.
+
+`app/data-strategies/event-logging.js`:
+
+```js
+import { EventLoggingStrategy } from '@orbit/coordinator';
+
+export default {
+  create() {
+    return new EventLoggingStrategy();
+  }
+};
+```
+
+Now when you reload your app you'll see a few log entries immediately:
+
+- store beforeQuery
+- store query
+
+When you add a message, you'll see a number of more entries:
+
+- store beforeUpdate
+- remote beforePush
+- remote transform
+- remote push
+- store transform
+- store update
+
+DETAILS ON WHAT EACH OF THESE MEANS SPECIFICALLY
+
+## Pull and Sync: Querying Data from the Server
+
+We're saving data to the server, but if you reload your app you'll notice we aren't loading the data yet; we simply query the in-memory store. What we want to do is, every time the in-memory store is queried, we want to query the remote store first.
+
+`app/data-strategies/remote-request.js`:
+
+```js
+import { RequestStrategy } from '@orbit/coordinator';
+
+export default {
+  create() {
+    return new RequestStrategy({
+      name: 'remote-request',
+
+      source: 'store',
+      on: 'beforeQuery',
+
+      target: 'remote',
+      action: 'pull',
+
+      blocking: true,
+    });
+  }
+};
+```
+
+This says that before the local store is queried, pull from the remote store.
+
+If you reload your app, this isn't enough to get the data load working. We also need to sync the data. The data is pulled down from the server, but it's left in limbo; it isn't transferred into our store. CONFIRM THIS EXPLANATION IS CORRECT. We need a sync strategy to do that:
```

Comment:
Is this explanation correct?

Reply:
Yes, sync strategies are used to "sync" every transform observed on one source into a target. If you `pull` data from a remote source, the data needs to be sync'd with the store.